Fix bug related to free shipping not being applied correctly

jacquesporveau opened the pull request:
**Description**
<!--
  Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

  Please include screenshots in case of visual changes to the frontend or backend sections of Solidus.

  If needed you can reference another PR or issue here, e.g.:
  Ref #ISSUE
-->

Fixes #3664 

I would leave a more detailed PR description but the author of #3664 did a really good job in describing the problem at hand and this PR is just a solution for that.

I kind of took a bit of liberty in improving the spec quality. Let me know if you don't like that style or something like that and I can rework it if need be.

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

spaghetticode commented on core/spec/models/spree/promotion/actions/free_shipping_spec.rb:
```diff
@@ -31,12 +31,55 @@
+    context "when a number of the orders shipments already have free shipping" do
+      let(:shipments) { order.shipments }
+
+      context "all the shipments have free shipping" do
```
I think this context description would sound better by adding `when` at the beginning

spaghetticode commented on core/spec/models/spree/promotion/actions/free_shipping_spec.rb:
```diff
@@ -31,12 +31,55 @@
+        end
+      end
+
+      context "some of the shipments have free shipping" do
```
same as above, I'd add `when` at the beginning

jarednorman commented on core/app/models/spree/promotion/actions/free_shipping.rb:
```diff
@@ -8,20 +8,21 @@ def perform(payload = {})
-              label: label
-            )
-            true
+          created_adjustments = order.shipments.map do |shipment|
+            if !promotion_credit_exists?(shipment)
+              shipment.adjustments.create!(
+                order: shipment.order,
+                amount: compute_amount(shipment),
+                source: self,
+                promotion_code: promotion_code,
+                label: label
+              )
+            end
```
Up to you which you think is better, but I probably would have switched this to use `next` to avoid unnecessary code churn.

```suggestion
          created_adjustments = order.shipments.map do |shipment|
            next if promotion_credit_exists?(shipment)
            
            shipment.adjustments.create!(
              order: shipment.order,
              amount: compute_amount(shipment),
              source: self,
              promotion_code: promotion_code,
              label: label
            )
```

jacquesporveau replied:
👍

jacquesporveau commented:
Thanks gang. [I've addressed the feedback.](https://github.com/solidusio/solidus/compare/08c47bffa3e11610b6573cc8ac48ebed168ed1a1..caf8957a584b3e1d612432af15ff42b11c33e4b2)
